src/main/java/com/powsybl/openloadflow/ac/IncrementalTransformerVoltageControlOuterLoop.java

```diff
@@ -93,100 +105,204 @@ private static void updateAllowedDirection(ControllerContext controllerContext,
         }
     }
 
-    private void adjustWithOneController(LfBranch controllerBranch, LfBus controlledBus, ContextData contextData, int[] controllerBranchIndex,
-                                         DenseMatrix sensitivities, double diffV, MutableObject<OuterLoopStatus> status) {
-        // only one transformer controls a bus
-        var controllerContext = contextData.getControllersContexts().get(controllerBranch.getId());
-        Double targetDeadband = controllerBranch.getTransformerVoltageControlTargetDeadband().orElse(null);
-        if (checkTargetDeadband(targetDeadband, diffV)) {
-            double sensitivity = ((EquationTerm<AcVariableType, AcEquationType>) controlledBus.getCalculatedV())
+    static class SensitivityContext {
+
+        private final DenseMatrix sensitivities;
+
+        private final int[] controllerBranchIndex;
+
+        public SensitivityContext(LfNetwork network, List<LfBranch> controllerBranches,
+                                  EquationSystem<AcVariableType, AcEquationType> equationSystem,
+                                  JacobianMatrix<AcVariableType, AcEquationType> j) {
+            controllerBranchIndex = createControllerBranchIndex(network, controllerBranches);
+            sensitivities = calculateSensitivityValues(controllerBranches, controllerBranchIndex, equationSystem, j);
+        }
+
+        private static int[] createControllerBranchIndex(LfNetwork network, List<LfBranch> controllerBranches) {
+            int[] controllerBranchIndex = new int[network.getBranches().size()];
+            for (int i = 0; i < controllerBranches.size(); i++) {
+                LfBranch controllerBranch = controllerBranches.get(i);
+                controllerBranchIndex[controllerBranch.getNum()] = i;
+            }
+            return controllerBranchIndex;
+        }
+
+        private static DenseMatrix calculateSensitivityValues(List<LfBranch> controllerBranches, int[] controllerBranchIndex,
+                                                              EquationSystem<AcVariableType, AcEquationType> equationSystem,
+                                                              JacobianMatrix<AcVariableType, AcEquationType> j) {
+            DenseMatrix rhs = new DenseMatrix(equationSystem.getIndex().getSortedEquationsToSolve().size(), controllerBranches.size());
+            for (LfBranch controllerBranch : controllerBranches) {
+                equationSystem.getEquation(controllerBranch.getNum(), AcEquationType.BRANCH_TARGET_RHO1)
+                        .ifPresent(equation -> rhs.set(equation.getColumn(), controllerBranchIndex[controllerBranch.getNum()], 1d));
+            }
+            j.solveTransposed(rhs);
+            return rhs;
+        }
+
+        @SuppressWarnings("unchecked")
```

Comment:
Do we need to keep this suppress warnings ?

Reply:
This is the purpose of moving this small piece of code into its own method: to avoid the warning and declare the cast is sage even if the compiler is not able to see it.